### ISSUE TITLE
example tools

### ISSUE THREAD

**Issue by brentp** [brentp/hts-nim-tools#3]
It would be nice to have a suite of example tools to further show off the syntax and utility of hts-nim. 
Please leave a comment here with a description.

The strengths of hts-nim are speed, so things that are too slow in python would be good candidates. It has very good syntax for VCF and BAM.

**Comment by brentp** [comment]
a VCF QC tool that takes a multi-sample VCF and finds big contiguous chunks where a sample has all missing genotypes could indicate that a processing chunk (e.g. from a single-sample GVCF) was lost. This could output a plotly plot showing the rate of missingness across each chrom for each sample.

**Comment by jdidion** [comment]
VCF-VCF comparison, to determine concordance (given two orthogonal VCFs) or sens/spec (given a truth set).

**Comment by andreas-wilm** [comment]
A pileup engine! I'll admit that doesn't really count as a tool but would enable the implementation of many.

**Comment by brentp** [comment]
These are good ideas--though maybe more than bite-sized. :)
I'll be writing a pileup engine soonish. 
I like the idea of VCF-to-VCF comparison, but there are a lot of considerations for that -- does it do normalization, does it do each sample? etc? I'll have to thnk more about a common use-case

**Comment by dmckean** [comment]
Here's one that I ported over from a bcftools plugin I wrote a while ago. I'm sure there are a few things that I should re-optimize for nim.

[tabulate.nim](https://gist.github.com/dmckean/1fb9455c7cbcb8703f461f8e2a819025) bins and tabulates integer data from the genotype column in a VCF, like DP, MQ, etc.

See also [tabulate.c](https://gist.github.com/dmckean/9b641956fde4bd710e0be559d701f6fd), the bcftools plugin.

**Comment by brainstorm** [comment]
A [`sambamba slice`](https://lomereiter.github.io/sambamba/docs/sambamba-slice.html), pretty please? :D

**Comment by brentp** [comment]
@brainstorm what would it do that sambamba slice does not? 

I guess there could be an API-ish way to access uncompressed blocks.

**Comment by brainstorm** [comment]
@brentp Sambamba is in a delicate spot when it comes to release engineering/CI/CD and overall reliability right now. `slice` is on 0.6.7, unreleasable for several platforms automatically in bioconda at the moment of writing this because it has several underlying issues (TL;DR: rabbit hole of fixes/updates):

https://twitter.com/braincode/status/1017216491265986560

I just wanted to have something functional/deployable to drop into our pipes ;)

**Comment by brentp** [comment]
I'll have a look. Not sure if htslib even exposes this--I think sambamba could do it because it has its own wrappers.

**Comment by dmckean** [comment]
@brainstorm I was crawling through source for `bcftools concat --naive` when I remembered this thread. 

[bcftools/vcfconcat.c::naive_concat](https://github.com/samtools/bcftools/blob/69711c976fa098537cc28858313cc53fea36a126/vcfconcat.c#L681_L689)  checks each BGZF block header ( at least to get the block size) and copies each block w/out decompressing. EOF blocks for each file are omitted, a single EOF block is placed at close. 

Bcftools' naive concatenation trusts that all input VCFs have compatible headers and writes out the first file's VCF header without checks. Naturally all input must be gzipped, but all input files must either be BCF or VCF.

@brentp I don't think that htslib provides an API for this beyond `bgzf_raw_read`, `bgzf_raw_write`, and other functions in `htslib/bgzf.h` and `htslib/hfile.h`, at least for the block IO. Do you thinks those two functions would be useful enough to be worth adding to `hts_concat.h` and `bgzf.nim`?

I'm not certain if htslib exposes block positions in the result from a `hts_itr_query` call, but that function calls a lot of helpers that may do most of the legwork for helping get a list of blocks for a queried region.

**Comment by brentp** [comment]
yes, if those are in the exposed api (in the header) files, then they can be added to hts_concat.h, if you or someone else wants to write an example or an hts-nim api on how to use the raw_read/write to get uncompressed blocks from a given region then it would be a great addtion to bgzf.nim.

**Comment by danielecook** [comment]
Some other bigger ideas but I may try to tackle a few:

1. Replacement for picard MarkDuplicates. This is a huge bottleneck for our pipeline. I may try to tackle.
2. `window` - generates regions from a sam / vcf for piping parallelizing in programs (e.g. parallelizing variant calling by chrom or region). I'm planning on adding this to seq-collection. Very useful with GNU-parallel.
3. Extracting reads that map to different contigs.
4. Extract read pairs that overlapr; Could you use this to estimate errors/artifacts?
